src/controller/CHIPDeviceController.cpp

```diff
@@ -1580,10 +1580,88 @@ void DeviceCommissioner::OnDeviceConnectionFailureFn(void * context, PeerId peer
 void DeviceCommissioner::SetupCluster(ClusterBase & base, DeviceProxy * proxy, EndpointId endpoint,
                                       Optional<System::Clock::Timeout> timeout)
 {
-    base.Associate(proxy, 0);
+    base.Associate(proxy, endpoint);
     base.SetCommandTimeout(timeout);
 }
 
+void DescriptorClusterPartsCallback(void * context, const chip::app::DataModel::DecodableList<chip::EndpointId> & data)
+{
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+    EndpointParts parts;
+    auto iter = data.begin();
+    while (iter.Next() && parts.numEndpoints < parts.kMaxEndpoints)
+    {
+        parts.endpoints[parts.numEndpoints++] = iter.GetValue();
+    }
+    CommissioningDelegate::CommissioningReport report;
+    report.Set<EndpointParts>(parts);
+    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
+}
+
+void DescriptorClusterServerCallback(void * context, const chip::app::DataModel::DecodableList<chip::ClusterId> & data)
+{
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+
+    bool hasGeneralCommissioning   = false;
+    bool hasNetworkCommissioning   = false;
+    bool hasBasic                  = false;
+    bool hasOperationalCredentials = false;
+
+    auto iter = data.begin();
+    while (iter.Next())
+    {
+        switch (iter.GetValue())
+        {
+        case app::Clusters::GeneralCommissioning::Id:
+            hasGeneralCommissioning = true;
+            break;
+        case app::Clusters::NetworkCommissioning::Id:
+            hasNetworkCommissioning = true;
+            break;
+        case app::Clusters::Basic::Id:
+            hasBasic = true;
+            break;
+        case app::Clusters::OperationalCredentials::Id:
+            hasOperationalCredentials = true;
+            break;
+        }
+    }
+    bool isCommissionable = hasGeneralCommissioning && hasBasic && hasOperationalCredentials;
```

Comment:
Need to check iter status?

Reply:
Checked by Next()

Reply:
See above about malformed data handling.

Reply:
added